compiler/src/iree/compiler/Codegen/LLVMGPU/ROCDLConfigureBufferInstructions.cpp

```diff
@@ -5,154 +5,214 @@
   }
-  // Note: this is a bit conservative, in that it will traverse all the
-  // arguments to a util.assume.int that isn't the immediate parent of val.
-  [[maybe_unused]] LogicalResult result =
-      getBackwardSlice(arg, &dependencies, opts);
-  assert(result.succeeded());
-  return llvm::all_of(dependencies, [&](Operation *op) {
-    if (matchPattern(op, m_Constant()))
-      return true;
-    if (isa<IREE::HAL::InterfaceConstantLoadOp, IREE::Util::AssumeIntOp>(op)) {
-      return true;
-    }
-    if (isa_and_nonnull<arith::ArithDialect>(op->getDialect())) {
-      return true;
-    }
-    return false;
-  });
+  if (auto vectorType = dyn_cast<VectorType>(type)) {
+    return vectorType.getNumElements() *
+           getTypeBitWidth(vectorType.getElementType());
+  }
+  // Over approximate the size of index to not require using the data layout.
+  if (type.isIndex()) {
+    return 8u;
+  }
+  // If it's not an int or float return 0.
+  return type.isIntOrFloat() ? type.getIntOrFloatBitWidth() : 0;
 }
 
-/// Return the maximum value that has been `util.assume.int`'d about this value
-/// if there is one.
-/// TODO: it'd be nice to be able to run the IntRangeAnalysis just up to the
-/// value in question, but we don't have that, so we approximate it.
-static std::optional<int64_t> getDynamicSizeMax(Value size) {
-  size = stripIntegerCasts(size);
-  // Special case for constants that're still dynamic.
-  APInt constVal;
-  if (matchPattern(size, m_ConstantInt(&constVal))) {
-    return constVal.getZExtValue();
+//===----------------------------------------------------------------------===//
+// ROCDLConfigureBufferInstructionsPass
+//===----------------------------------------------------------------------===//
+
+/// Returns whether the binding op satisfies the requirement of being less than
+/// 2GB in size.
+static bool
+isValidTensorSpan(IREE::HAL::InterfaceBindingSubspanOp op,
+                  llvm::function_ref<int64_t(OpFoldResult)> getMax) {
+  RankedTensorType tensorTy =
+      cast<IREE::TensorExt::DispatchTensorType>(op.getType())
+          .asRankedTensorType();
+  ArrayRef<int64_t> shape = tensorTy.getShape();
+
+  // Return immediately if the tensor is trivial.
+  if (shape.empty()) {
+    return true;
   }
-  auto assumeOp = size.getDefiningOp<IREE::Util::AssumeIntOp>();
-  if (!assumeOp)
-    return std::nullopt;
-  std::optional<int64_t> maybeMax =
-      assumeOp.getUnionedUnsignedRange(cast<OpResult>(size).getResultNumber())
-          .second;
-  return maybeMax;
-}
 
-static std::optional<int64_t>
-getSpannedBytes(IREE::HAL::InterfaceBindingSubspanOp binding) {
-  int64_t maxNumElems = 1;
-  ShapedType resultTy = dyn_cast<ShapedType>(binding.getType());
-  if (auto tensorType =
-          dyn_cast<IREE::TensorExt::DispatchTensorType>(binding.getType())) {
-    resultTy = tensorType.asRankedTensorType();
+  // Compute the max allowed size.
+  unsigned byteSize = getTypeBitWidth(tensorTy.getElementType());
```

Comment:
This is in units of bits, not bytes, and you want 2 GB times 8 therefore, below

Reply:
I'd not sure this works right for `i4` and friends

Can we use units of bits as long as possible?

Reply:
Yes, but handling fp4 tensors and such correctly requires doing your math in bits and then doing a ceiling-divide by 8 at the very end